### ISSUE TITLE
Create an intent-in compiler option

### ISSUE THREAD

**Issue by Beliavsky** [lfortran/lfortran#2328]
Reading the blog post [C++ papercuts](https://www.thecodedmessage.com/posts/c++-papercuts/) complaining that "const is not the default" in C++ gave me the idea that it would be ideal if arguments of Fortran procedures were `intent(in)` by default, with the user having to explicitly declare them `intent(out)` or `intent(in out)` is that's what she wanted. We can't break Fortran with such an incompatible change, but I suggest that LFortran have an intent-in option to give such behavior.

I understand the LFortran team is focused on implementing Fortran rather than non-standard enhancements but thought I would mention this idea before I forget it. Maybe I could write a Python script that annotates all Fortran procedure arguments as `intent(in)` unless they are declared otherwise.

**Comment by certik** [comment]
In LPython the default is `intent(in)` if you don't specify anything. You have to say Out or InOut for the other two.

In LFortran we have two options:

* Either just enforce "intent" to be present at all times by default, unless a compiler option is present
* If intent is not specified, it will be "intent(in)", but I don't like this option, it feels fragile

I think the first option is the obvious one. LFortran by default is definitely non-standard, the requirement to type all variables is the most visible "deviation". 

So I think we should just require the user to always provide "intent". All my codes I think will pass. For older codes we provide a compiler option to disable this "pedantic" default.

Problem solved.

**Comment by Beliavsky** [comment]
I see you have thought about this. A benefit of your 2nd LFortran option is that it would take less work to make an old code compile, since the arguments that should be intent(in) could be left unannotated. I agree that all arguments should have intents specified, but if the compiler is too demanding, people may just disable the pedantic options.

**Comment by certik** [comment]
The problem is that making unspecified equal to `intent(in)` will break old codes. So either way this default option would have to be turned off for non-conforming codes.

**Comment by dacarnazzola** [comment]
I like option 1:
>In LFortran we have two options:
>
>    Either just enforce "intent" to be present at all times by default, unless a compiler option is present

Encouraging good coding practice is beneficial to everyone. Some escape hatch to turn it off and compile like the wild west is fine. Lack of `intent` already has weird enough behavior without adding `implicit intent(in)`. Fortran is a rather verbose language by today's standards, at least we can encourage _explicit_ code where possible. This option would increase clarity, rather than starting down the path of creating a new set of implicit defaults.

**Comment by certik** [comment]
I agree, it's good to be demanding. Unlike Rust, we can turn it off.
